### PR TITLE
Implement metadata v15

### LINKED ISSUE #192: implement Metadata v15
The current supported version is `v14`. Implement `v15`.
* [Specification](https://spec.polkadot.network/sect-metadata)
* [Type](https://github.com/paritytech/frame-metadata/blob/main/frame-metadata/src/v15.rs#L48)
* [Imported in Substrate](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/metadata-ir/src/v15.rs#L34)
* `Metadata_metadata` should return `v15`. You can extract `RuntimeMetadataV14` into another file and create a separate file for `RuntimeMetadataV15`. Do not remove `v14` as it will be used in #193 .

- [ ] Add unit and integration tests, which assert that the metadata contains the newly added properties.

### LINKED ISSUE #193: implement runtime metadata api - `Metadata_metadata_at_version` and `Metadata_metadata_versions`
There are two new exported runtime metadata api functions:
* [Metadata_metadata_at_version](https://spec.polkadot.network/chap-runtime-api#sect-rte-metadata-metadata-at-version)
* [Metadata_metadata_versions](https://spec.polkadot.network/chap-runtime-api#sect-rte-metadata-metadata-versions)

Add support for `v14` and `v15`.

### PR DESCRIPTION
**Pull request by nicksinch**
**Detailed description**:
Adds support for **Metadata V15**:

- Adds new structs required by the v15 spec
- Adds new metadata types in some of the frame modules as well as in the metadata api
- Two more runtime functions exported: **Metadata_at_version(version)** **and Metadata_versions()**
-  Leaves the default metadata version returned to be 14
- Adjusts the tests to expect a default metadata version 14

**Which issue(s) this PR fixes**:
Fixes #192 , #193 

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: execution/extrinsic/runtime_extrinsic.go [hunk 1]
```diff
@@ -246,6 +329,33 @@ func (re runtimeExtrinsic) runtimeEvent(variants sc.Sequence[sc.Option[primitive
...
+			primitives.NewMetadataDefinitionVariant(
+				"Grandpa",
+				sc.Sequence[primitives.MetadataTypeDefinitionField]{
+					primitives.NewMetadataTypeDefinitionField(metadata.TypesGrandpaErrors),
+				},
+				runtimeGrandpaErrIdx,
+				""),
+		}))
+}
+
```
Comment (failfmi) [inline execution/extrinsic/runtime_extrinsic.go]: Can we refactor it to be the same as `runtimeEvent` and `runtimeCall`?

File: runtime/runtime_test.go [hunk 2]
```diff
@@ -20,9 +20,9 @@ import (
 )
 
 const POLKADOT_RUNTIME = "../build/polkadot_runtime-v9400.compact.compressed.wasm"
-const NODE_TEMPLATE_RUNTIME = "../build/node_template_runtime.wasm"
+const NODE_TEMPLATE_RUNTIME = "../build/node_template_v15.wasm"
 
-// const WASM_RUNTIME = "../build/node_template_runtime.wasm"
+// const WASM_RUNTIME = "../build/node_template_v15.wasm"
```
Comment (failfmi) [inline runtime/runtime_test.go]: nit: revert this

File: frame/system/module.go [hunk 3]
```diff
@@ -25,6 +25,36 @@ const (
...
+const (
+	CannotLookupIndex sc.U8 = iota
+	NoUnsignedValidatorIndex
+	CustomUnknownIndex
+)
+
+const (
+	InvalidTransactionIndex sc.U8 = iota
+	UnknownTransactionIndex
+)
```
Comment (failfmi) [inline frame/system/module.go]: You have these in `types`

File: frame/system/module.go [hunk 4]
```diff
@@ -25,6 +25,36 @@ const (
...
+
+const (
+	ValidTransactionIndex sc.U8 = iota
+	TransactionErrIndex
+)
+
```
Comment (failfmi) [inline frame/system/module.go]: You have these in `types`

File: frame/system/module.go [hunk 5]
```diff
@@ -25,6 +25,36 @@ const (
...
+	BadMandatoryIndex
+	MandatoryValidationIndex
+	BadSignerIndex
+)
+
+const (
+	CannotLookupIndex sc.U8 = iota
+	NoUnsignedValidatorIndex
+	CustomUnknownIndex
+)
```
Comment (failfmi) [inline frame/system/module.go]: You have these in `types`

File: api/metadata/module.go [hunk 6]
```diff
@@ -88,11 +163,13 @@ func primitiveTypes() sc.Sequence[primitives.MetadataType] {
...
+		primitives.NewMetadataType(metadata.TypesFixedSequence8U8, "[8]byte", primitives.NewMetadataTypeDefinitionFixedSequence(8, sc.ToCompact(metadata.PrimitiveTypesU8))),
 		primitives.NewMetadataType(metadata.TypesFixedSequence20U8, "[20]byte", primitives.NewMetadataTypeDefinitionFixedSequence(20, sc.ToCompact(metadata.PrimitiveTypesU8))),
 		primitives.NewMetadataType(metadata.TypesFixedSequence32U8, "[32]byte", primitives.NewMetadataTypeDefinitionFixedSequence(32, sc.ToCompact(metadata.PrimitiveTypesU8))),
 		primitives.NewMetadataType(metadata.TypesFixedSequence64U8, "[64]byte", primitives.NewMetadataTypeDefinitionFixedSequence(64, sc.ToCompact(metadata.PrimitiveTypesU8))),
 		primitives.NewMetadataType(metadata.TypesFixedSequence65U8, "[65]byte", primitives.NewMetadataTypeDefinitionFixedSequence(65, sc.ToCompact(metadata.PrimitiveTypesU8))),
 		primitives.NewMetadataType(metadata.TypesSequenceU8, "[]byte", primitives.NewMetadataTypeDefinitionSequence(sc.ToCompact(metadata.PrimitiveTypesU8))),
+		primitives.NewMetadataType(metadata.TypesSequenceU32, "[]byte", primitives.NewMetadataTypeDefinitionSequence(sc.ToCompact(metadata.PrimitiveTypesU32))),
```
Comment (failfmi) [inline api/metadata/module.go]: You can rename `[]byte` -> `[]uint32`

Comment (nicksinch) [reply api/metadata/module.go]: Done.

File: frame/system/module.go (hunk 5 above)
Comment (nicksinch) [reply frame/system/module.go]: Right, removed from here.

File: frame/system/module.go (hunk 4 above)
Comment (nicksinch) [reply frame/system/module.go]: I don't see those 2 but I've removed them from here and defined them similar to others in primitives/types/transaction_validity_result.go

File: frame/system/module.go (hunk 3 above)
Comment (nicksinch) [reply frame/system/module.go]: Removed from here.

File: runtime/runtime_test.go (hunk 2 above)
Comment (nicksinch) [reply runtime/runtime_test.go]: Done.

File: frame/system/module.go (hunk 4 above)
Comment (failfmi) [reply frame/system/module.go]: Yeah, you can update `DecodeTransactionValidityResult` and `Encode` to use them. Could you rename them to `TransactionValidityResultValid` and `TransactionValidityResultError`?

File: execution/extrinsic/runtime_extrinsic.go (hunk 1 above)
Comment (nicksinch) [reply execution/extrinsic/runtime_extrinsic.go]: runtimeCall and runtimeEvent use a sequence of CallDef and EventDef respectively to build the type. So, to be the same as them, you mean to add ErrorDef in the MetadataModule struct and do it analogically ?

Comment (nicksinch) [reply execution/extrinsic/runtime_extrinsic.go]: Done.

File: frame/system/module.go (hunk 4 above)
Comment (nicksinch) [reply frame/system/module.go]: Done.

File: runtime/runtime.go [hunk 7]
```diff
@@ -312,6 +312,20 @@ func Metadata(_, _ int32) int64 {
 		Metadata()
 }
 
+//go:export Metadata_at_version
```
Comment (failfmi) [inline runtime/runtime.go]: Function name must be `Metadata_metadata_at_version`
https://spec.polkadot.network/chap-runtime-api#sect-rte-metadata-metadata-at-version

File: runtime/runtime.go [hunk 8]
```diff
@@ -312,6 +312,20 @@ func Metadata(_, _ int32) int64 {
...
+func MetadataAtVersion(dataPtr int32, dataLen int32) int64 {
+	return runtimeApi().
+		Module(metadata.ApiModuleName).(metadata.Module).
+		MetadataAtVersion(dataPtr, dataLen)
+}
+
+//go:export Metadata_versions
```
Comment (failfmi) [inline runtime/runtime.go]: Function name must be `Metadata_metadata_versions`
https://spec.polkadot.network/chap-runtime-api#sect-rte-metadata-metadata-versions

File: api/metadata/module.go [hunk 9]
```diff
@@ -61,7 +81,62 @@ func (m Module) buildMetadata() primitives.Metadata {
...
+	return primitives.Metadata{
+		Version: primitives.MetadataVersion14,
+		DataV14: runtimeV14Metadata,
+	}
+}
+
+// MetadataAtVersion returns the metadata of a specific version of the runtime passed as argument.
+// Returns a pointer-size of the SCALE-encoded metadata of the runtime.
+// [Specification](https://spec.polkadot.network/chap-runtime-api#sect-rte-metadata-metadata)
+func (m Module) MetadataAtVersion(dataPtr int32, dataLen int32) int64 {
```
Comment (failfmi) [inline api/metadata/module.go:94]: This method returns an option of metadata, as some versions might not be supported.

**Review by failfmi — CHANGES_REQUESTED** [review]

File: tinygo
Comment (radkomih) [inline tinygo:1]: You can still use the latest Tinygo changes, just do not commit them yet, at least for now. 
I'll take a look at the fix and why the tests are failing

File: runtime/runtime.go (hunk 7 above)
Comment (nicksinch) [reply runtime/runtime.go]: Fixed

File: runtime/runtime.go (hunk 8 above)
Comment (nicksinch) [reply runtime/runtime.go]: Fixed

File: api/metadata/module.go (hunk 9 above)
Comment (nicksinch) [reply api/metadata/module.go:94]: Done.

File: constants/metadata/metadata.go [hunk 10]
```diff
@@ -123,4 +132,42 @@ const (
...
+
+	TypesBlock
+	TypeOptionOpaqueMetadata
+	TypeOptionInclusionFee
+	TypesOptionTupleSequenceU8KeyTypeId
+
+	TypesResultEmptyTuple
+
+	TypesGrandpaAppPublic
+	TypesTupleGrandaAppPublicU64
```
Comment (failfmi) [inline constants/metadata/metadata.go]: nit: typo

File: execution/extrinsic/runtime_extrinsic.go [hunk 11]
```diff
@@ -8,6 +8,12 @@ import (
 	primitives "github.com/LimeChain/gosemble/primitives/types"
 )
 
+const (
+	runtimeSystemErrIdx sc.U8 = iota
+	runtimeBalancesErrIdx
+	runtimeGrandpaErrIdx
+)
+
```
Comment (failfmi) [inline execution/extrinsic/runtime_extrinsic.go]: nit: unused

**Review by failfmi — DISMISSED** [review]

**Review by radkomih — DISMISSED** [review]

**Review by failfmi — APPROVED** [review]

**Review by radkomih — APPROVED** [review]

**Pull request merged**